### PR TITLE
[Refactor] refactor encoding interface

### PR DESCRIPTION
**Pull request by ly015**
## Motivation

Refactor the encoding interface for better extendibility and easier configuration.

- **Codec**: The `encode()` method will return a dict that specifies the key of the encoded item.
- **GenerateTarget**: Directly update the encoded items (keys and values) into results. The argument `target_type` is no longer needed. Support 3 encoding modes:
  - Single (single encoder): Encoded items will be updated into results
  - Multilevel (multiple encoders; `multilevel=True`): All encoders must have the same keys of encoded items. Multilevel encoded items will be stored in lists and handled later in `PackPoseInputs`
  - Combined (multiple encoders; `multilevel=False`): All encoders must have no overlapping keys (except 'keypoint_weights') of encoded items. All encoded items will be updated into results. Multiple 'keypoint_weights' will be multiplied together and updated into results.
- **PackPoseInput**: Add support for multilevel labels.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/datasets/transforms/common_transforms.py [hunk 1]
```diff
@@ -873,58 +874,136 @@ class GenerateTarget(BaseTransform):
...
+                        'Encoded items from all encoders must have the same '
+                        'keys if ``multilevel==True``.')
+
+                encoded = {
+                    k: [_encoded[k] for _encoded in encoded_list]
+                    for k in keys
+                }
+
+            else:
+                # For combined encoding, the encoded items from different
```
Comment (liqikai9) [inline mmpose/datasets/transforms/common_transforms.py:968]: Is it necessary to handle the case that we want to use different encoded `keypoint_weights`  to calculate different losses for different codecs? Should we keep all the `keypoint_weights` for possible usage?

File: mmpose/datasets/transforms/common_transforms.py [hunk 2]
```diff
@@ -873,58 +874,136 @@ class GenerateTarget(BaseTransform):
...
-                 target_type: str,
+                 target_type: Optional[str] = None,
+                 multilevel: bool = False,
                  use_dataset_keypoint_weights: bool = False) -> None:
         super().__init__()
+
+        if target_type is not None:
+            warnings.warn(
+                'The argument `target_type` in GenerateTarget is deprecated.'
+                'The target type and encoded keys will be directly')
```
Comment (liqikai9) [inline mmpose/datasets/transforms/common_transforms.py:901]: Seems like miss some message here: wii be directly updated into the results dict.

File: mmpose/datasets/transforms/common_transforms.py [hunk 3]
```diff
@@ -873,58 +874,136 @@ class GenerateTarget(BaseTransform):
...
-                labels and will be stored in ``results['keypoint_label']``
-            - ``'keypoint_xy_label'``: The encoed should be instance-level
-                labels in x-axis and y-axis respectively. They will be stored
-                in ``results['keypoint_x_label']`` and
-                ``results['keypoint_y_label']``
-            - ``'heatmap+keypoint_label'``: The encoded should be heatmaps and
-                keypoint_labels, will be stored in ``results['heatmaps']``
-                and ``results['keypoint_label']``
+        target_type (str, deprecated): This argument is deprecated and has no
+            effect
```
Comment (liqikai9) [inline mmpose/datasets/transforms/common_transforms.py:886]: missing description about `multilevel` in the docstring

File: mmpose/datasets/transforms/common_transforms.py [hunk 4]
```diff
@@ -873,58 +874,136 @@ class GenerateTarget(BaseTransform):
...
             self.encoder = KEYPOINT_CODECS.build(self.encoder_cfg)
 
     def transform(self, results: Dict) -> Optional[dict]:
+        """The transform function of :class:`GenerateTarget`.
+
+        See ``transform()`` method of :class:`BaseTransform` for details.
+        """
+
+        if results.get('transformed_keypoints', None) is not None:
+            # use keypoints transformed by TopdownAffine
```
Comment (liqikai9) [inline mmpose/datasets/transforms/common_transforms.py:923]: `TopdownAffine` may be inaccurate here, how about `affine transformation`

File: mmpose/datasets/transforms/formatting.py [hunk 5]
```diff
@@ -88,13 +88,19 @@ class PackPoseInputs(BaseTransform):
...
     label_mapping_table = {
         'keypoint_labels': 'keypoint_labels',
         'keypoint_x_labels': 'keypoint_x_labels',
         'keypoint_y_labels': 'keypoint_y_labels',
         'keypoint_weights': 'keypoint_weights'
     }
 
+    # items in `field_mapping_table` will be packed into
+    # PoseDataSample.gt_fields and converted to Tensor. These items will be
+    # used for computing lossses
```
Comment (Ben-Louis) [inline mmpose/datasets/transforms/formatting.py:103]: lossses -> losses

**Review by Tau-J — APPROVED** [review]

File: docs/en/user_guides/configs.md [hunk 6]
```diff
@@ -174,7 +174,7 @@ train_pipeline = [ # data aug in training
     dict(type='TopdownAffine', input_size=codec['input_size']), # update inputs via transform matrix
     dict(
         type='GenerateTarget', # generate targets via transformed inputs
-        target_type='heatmap', # typeof targets
+        # typeof targets
```
Comment (Ben-Louis) [inline docs/en/user_guides/configs.md:177]: This comment seems deprecated

File: docs/zh_cn/user_guides/configs.md [hunk 7]
```diff
@@ -179,7 +179,7 @@ train_pipeline = [ # 训练时数据增强
     dict(type='TopdownAffine', input_size=codec['input_size']), # 根据变换矩阵更新目标数据
     dict(
         type='GenerateTarget', # 根据目标数据生成监督信息
-        target_type='heatmap', # 监督信息类型
+        # 监督信息类型
```
Comment (liqikai9) [inline docs/zh_cn/user_guides/configs.md:182]: This line may also be redundant like the last comment

**Pull request merged**